src/lib/core/sample_sort.c

```diff
@@ -0,0 +1,513 @@
+ *
+ * Additionally we check if data is already sorted before applying main algo.
+ *
+ * See also:
+ * [1] https://en.wikipedia.org/wiki/Samplesort
+ * [2] Super Scalar Sample Sort, Peter Sanders and Sebastian Winkel
+ *
+ * Although this implementation does not use superscalar approach to map
+ * elements to buckets as in [2] and use usual binary search.
+ */
+void
+sample_cosort_mt(void *data, size_t data_size, size_t elem_size,
+		 compare_arg_f cmp, void *cmp_arg, int threads_num)
+{
+	struct sample_sort sort;
+	double b, e;
+
+	say_verbose("start sort, data size: %zd, elem size: %zd, threads: %d\n",
+		    data_size, elem_size, threads_num);
+	if (data_size < NOSPAWN_SIZE_THESHOLD) {
+		say_verbose("data size is less then threshold %d,"
+			    " sort in caller thread", NOSPAWN_SIZE_THESHOLD);
+		qsort_arg_st(data, data_size, elem_size, cmp, cmp_arg);
+		return;
+	}
+
+	/* Upper limit is because oracles are chars. */
+	assert(threads_num > 0 && threads_num <= SAMPLE_SORT_THREADS_MAX);
+
+	sort.data = data;
+	sort.data_size = data_size;
+	sort.elem_size = elem_size;
+	sort.cmp = cmp;
+	sort.cmp_arg = cmp_arg;
+	sort.threads_num = threads_num;
+
+	if (threads_num == 1) {
+		say_verbose("sorting thread number is 1,"
+			    " fallback to qsort");
+		struct cord cord;
+		if (cord_costart(&cord, "sort.worker.0", sample_sort_sort_all,
+				 &sort) != 0) {
+			diag_log();
+			panic("cord_start failed");
+		}
+		if (cord_cojoin(&cord) != 0) {
+			diag_log();
+			panic("cord_cojoin failed");
+		}
+		return;
+	}
+
+	sort.oracles = xmalloc(data_size);
+	sort.buffer = xmalloc(data_size * elem_size);
+	sort.splitters = xmalloc((threads_num - 1) * elem_size);
+
+	size_t part_size = data_size / threads_num;
+	struct sample_sort_worker *workers =
+		xmalloc(threads_num * sizeof(struct sample_sort_worker));
+	bool presorted = true;
+	/* Required for presorted check on part borders. */
+	assert(part_size > 0);
+	for (int i = 0; i < threads_num; i++) {
+		struct sample_sort_worker *worker = &workers[i];
+
+		worker->sort = &sort;
+		worker->begin = i * part_size;
+		/*
+		 * Each thread takes equal share of data except for last
+		 * thread which takes extra `data_size % threads_num` elements
+		 * if `data_size` in not multiple of `threads_num`.
+		 */
+		if (i == threads_num - 1)
+			worker->end = data_size;
+		else
+			worker->end = worker->begin + part_size;
+
+		worker->bucket_hist = xcalloc(threads_num, sizeof(size_t));
+		worker->bucket_offs = xmalloc(threads_num * sizeof(size_t));
+
+		if (presorted && i < threads_num - 1 &&
+		    cmp(data + (worker->end - 1) * elem_size,
+			data + worker->end * elem_size, cmp_arg) > 0)
+			presorted = false;
+	}
+
+	if (presorted) {
+		sample_sort_corun_mt(sample_sort_check_presorted, workers,
+				     threads_num);
+		for (int i = 0; i < threads_num; i++) {
+			if (!workers[i].presorted) {
+				presorted = false;
+				break;
+			}
+		}
+		if (presorted) {
+			say_verbose("data is presorted\n");
```

Comment:
`\n` is redundant. Please remove from all messages.

Reply:
Fixed.